onelogin_rules/onelogin_brute_force_user_login.py

```diff
@@ -0,0 +1,51 @@
+from panther_oss_helpers import evaluate_threshold  # pylint: disable=import-error
+
+# TODO change to native thresholding once support is added
+THRESH = 10
+THRESH_TTL = 600  # 10 minutes
+
+
+def rule(event):
+
+    # filter events; event type 6 is a failed authentication
+    if event.get('event_type_id') != 6:
+        return False
+
+    # keep track of user authentication failures
+    # using the same username can trigger a brute force alert
+    if (evaluate_threshold(
+            '{}-OneLoginFailedCounterUsername'.format(
+                event.get('actor_user_id')),
+            THRESH,
+            THRESH_TTL,
+    )):
+        return True
+
+    # failed logins originating from a single ip,
+    # but not necessarily using the same username can trigger
+    # a brute force alert (e.g. password spraying)
+    if evaluate_threshold(
+            '{}-OneLoginFailedCounterIpAddress'.format(event.get('ipaddr')),
+            THRESH,
+            THRESH_TTL,
+    ):
+        return True
```

Comment:
Is there a way to do threshold based on different fields? E.g. threshold for usernames (brute force single user's password) as well as ip (single ip tries to login to many different accounts).  Or is the idea that the new way will trigger off of total number of failures, and it doesn't matter if its unique usernames, unique ips, etc?

Reply:
My understanding is that the thresholding is based on alerts with the same dedup string, but I'm not 100% sure.

So from my understanding, if the dedup function returns `event['ip_address']` and the dedup threshold is 10, then an alert will be fired if 10 events trigger with the same `event['ip_address']` value with a `dedup_period_minutes` window.

Reply:
> My understanding is that the thresholding is based on alerts with the same dedup string, but I'm not 100% sure.

This is correct

Reply:
With that in mind, I made this one rule into two: one for username dedup and one for ip address dedup